src/include/statistics/abstract_metric.h

```diff
@@ -4,48 +4,185 @@
+   * ensures this method does not return if the collecting thread still can
+   * write to the old raw data.
+   *
+   * @return a shared pointer to the old AbstractRawData
+   */
+  virtual std::shared_ptr<AbstractRawData> Swap() = 0;
+};
+
+/* Forward Declaration */
+template <typename DataType>
+class AbstractMetric;
+
+/**
+ * @brief Wraps around a pointer to an AbstractRawData to allow safe access.
+ *
+ * This class is always handed out by an AbstractMetric and would prevent an
+ * Aggregator from reading or deleting the AbstractRawData it holds. When the
+ * object goes out of scope, its destructor will unblock the aggregator. Access
+ * to its underlying pointer is always non-blocking.
+ *
+ * @tparam DataType the type of AbstractRawData this Wrapper holds
+ */
+template <typename DataType>
+class RawDataWrapper {
+  friend class AbstractMetric<DataType>;
 
-  const inline MetricType& GetType() const { return type_; }
+ public:
+  inline RawDataWrapper(RawDataWrapper &&other) = default;
 
-  virtual void Reset() = 0;
+  inline ~RawDataWrapper() { safe_ = true; }  // Unblock aggregator
 
-  virtual const std::string GetInfo() const = 0;
+  DISALLOW_COPY(RawDataWrapper);
 
-  virtual void Aggregate(AbstractMetric& source) = 0;
+  /**
+   * @return the underlying pointer
+   */
+  inline DataType *operator->() const { return ptr_; }
 
  private:
-  // The type this metric belongs to
-  MetricType type_;
+  /**
+   * Constructs a new Wrapper instance
+   * @param ptr the pointer it wraps around
+   * @param safe the boolean variable it uses to signal its lifetime
+   */
+  inline RawDataWrapper(DataType *ptr, bool &safe) : ptr_(ptr), safe_(safe) {}
+  DataType *ptr_;
+  bool &safe_;
 };
 
+/**
+ * @brief General purpose implementation to Metric that you should inherit from.
+ *
+ * This class implements the tricky Swap method and exposes an interface for
+ * children class. @see Metric for detail
+ *
+ * @tparam DataType the type of AbstractRawData this Metric holds
+ */
+template <typename DataType>
+class AbstractMetric : public Metric {
+ public:
+  AbstractMetric() : raw_data_(new DataType()) {}
+  /**
+   * @see Metric
+   *
+   * To ensure this method works as intended, be sure to use GetRawData() to
+   * access the underlying raw data
+   * @return a shared pointer to the old AbstractRawData
+   */
+  std::shared_ptr<AbstractRawData> Swap() override {
+    // After this point, the collector thread can not see old data on new
+    // events, but will still be able to write to it, if they loaded the
+    // pointer before this operation but haven't written to it yet.
+    DataType *old_data = raw_data_.exchange(new DataType());
+    // We will need to wait for last writer to finish before it's safe
+    // to start reading the content. It is okay to block since this
+    // method should only be called from the aggregator thread.
+    while (!safe_) _mm_pause();
```

Comment:
safe_ should be atomic? (volatile might work as well in this case, I guess).

-- Priyatham, Siva and Vamshi